drivers/uaol/uaol_intel_adsp.c

```diff
@@ -0,0 +1,861 @@
+	union UAOLCTL ctl;
+	uint64_t start_time;
+	uint64_t timeout = k_us_to_cyc_ceil64(UAOL_POWER_CHANGE_TIMEOUT_USEC);
+	uint32_t link_mask = BIT(dp->link);
+
+	ctl.full = sys_read32(UAOLCTL_ADDR(dp));
+	if ((ctl.part.cpa & link_mask) != (ctl.part.spa & link_mask)) {
+		return -EBUSY;
+	}
+	ctl.part.spa = (power) ? (ctl.part.spa | link_mask) : (ctl.part.spa & ~link_mask);
+	sys_write32(ctl.full, UAOLCTL_ADDR(dp));
+
+	start_time = sys_clock_cycle_get_64();
+	while (1) {
+		ctl.full = sys_read32(UAOLCTL_ADDR(dp));
+		if ((ctl.part.cpa & link_mask) == (ctl.part.spa & link_mask)) {
+			break;
+		}
+		if (sys_clock_cycle_get_64() - start_time > timeout) {
+			return -ETIMEDOUT;
+		}
+		k_busy_wait(1);
+	}
+
+	return 0;
+}
+
+/*
+ * Read HW capabilities for UAOL link.
+ */
+static void uaol_intel_adsp_read_capabilities(const struct device *dev,
+					      struct uaol_capabilities *caps)
+{
+	struct uaol_intel_adsp_data *dp = dev->data;
+	union UAOLxPCMSCAP pcmscap;
+	union UAOLxOPC opc;
+	union UAOLxIPC ipc;
+
+	pcmscap.full = sys_read16(UAOLxPCMSCAP_ADDR(dp));
+	opc.full = sys_read16(UAOLxOPC_ADDR(dp));
+	ipc.full = sys_read16(UAOLxIPC_ADDR(dp));
+
+	caps->input_streams = pcmscap.part.iss;
+	caps->output_streams = pcmscap.part.oss;
+	caps->bidirectional_streams = pcmscap.part.bss;
+	caps->max_tx_fifo_size = opc.part.opc;
+	caps->max_rx_fifo_size = ipc.part.ipc;
+}
+
+/*
+ * Program the target destination ID of the xHCI Controller for UAOL link.
+ */
+static void uaol_intel_adsp_program_bdf(const struct device *dev, uint8_t bus,
+					uint8_t device, uint8_t function)
+{
+	struct uaol_intel_adsp_data *dp = dev->data;
+	union UAOLxTBDF tbdf = {0};
+
+	tbdf.part.busn = bus;
+	tbdf.part.devn = device;
+	tbdf.part.fncn = function;
+	sys_write16(tbdf.full, UAOLxTBDF_ADDR(dp));
+}
+
+/*
+ * Program the FIFO Start Address Offset for UAOL stream.
+ */
+static void uaol_intel_adsp_program_fifo_start_offset(const struct device *dev, int stream,
+						      uint16_t fifo_start_offset)
+{
+	struct uaol_intel_adsp_data *dp = dev->data;
+
+	sys_write16(fifo_start_offset, UAOLxPCMSyFSA_ADDR(dp, stream));
+}
+
+/*
+ * Program channel mapping for UAOL stream.
+ */
+static void uaol_intel_adsp_program_channel_map(const struct device *dev, int stream,
+						uint16_t channel_map)
+{
+	struct uaol_intel_adsp_data *dp = dev->data;
+
+	sys_write16(channel_map, UAOLxPCMSyCM_ADDR(dp, stream));
+}
+
+/*
+ * Send xHCI message using the Immediate Command mechanism for UAOL link.
+ */
+static int uaol_intel_adsp_send_xhci_msg(const struct device *dev, uint32_t *msg, size_t msg_size)
+{
+	struct uaol_intel_adsp_data *dp = dev->data;
+	union UAOLxIC ic;
+	uint64_t start_time;
+	uint64_t timeout = k_us_to_cyc_ceil64(XHCI_MSG_TIMEOUT_USEC);
+	size_t dword_idx = 0;
+	size_t dword_count = msg_size / sizeof(msg[0]);
+	int i;
+
+	do {
+		/* write a 4x32-bit message chunk */
+		for (i = 0; i < 4 && dword_idx < dword_count; i++) {
+			sys_write32(msg[dword_idx++], UAOLxICPy_ADDR(dp, i));
```

Comment:
FWIW, this also has an uninitialized data bug if msg_size is not an integral multiple of 4.  It will write 0-3 bytes of data **after** the end of the passed msg array.

My guess is that that "should never happen", in which case __ASSERT() please.  Or change the API to take an number of words.  Or require that msg[] always be 16 bytes, etc...

Reply:
Added `BUIL_ASSERT()` to ensure that memory objects being passed as `msg_buf` to `send_xhci_msg()` and `receive_xhci_msg()` functions are the whole multiple of 4 in size.